### ISSUE TITLE
Allow multiple working copies for the same resource

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#92037]
## Problem
Using the custom editor API, I think we have a figure our how to allow there to be multiple working copies for a single resource. Consider a few example scenarios:

**Split custom editor**
- Open an image backed by a custom editor
- Split the image
- Reopen the second image with VS Code's standard editor

This causes two working copies for the same resource to be registered

**Text Document**
- Have an extension that opens a `TextDocument` for a given text resource
- Then open that same resource using a `CustomEditor` (not `CustomTextEditor`)

This causes two working copies for the same resource to be registered

## Proposal
I think we need to rethink our working copy service to support cases like the above. Some potential ideas:

- Allow multiple working copies to be registered for a single resource
- Associate editor inputs to be associated with specific working copies instead of just with resources

**Comment by bpasero** [comment]
@mjbvz I am happy to revert https://github.com/microsoft/vscode/commit/e8f005af1955ccf23abc5c5e11a6179dea017412 which basically now throws if you register a working copy with the same resource once we find a way to get backups right: the reason I pushed for this change is that our backup service is not fit for multiple working copies on the same resource. Here is a data loss scenario:

* user opens a normal text editor and a custom editor (which does not operate on the text document but introduces its own working copy) which has the same resource
* both editors are dirty
* user quits VSCode
* our backup logic kicks in and writes a backup for each dirty editor
* both editors use the same resource, so the one editor will overwrite the backup of the other

At this point the data from one editor is guaranteed to be lost because there is no guarantee that either the custom editor or the text editor write in a format that is compatible with the other.

Long story short: backup service operates on resources and as such working copies need to be unique per resource to prevent data loss.

**Comment by bpasero** [comment]
@mjbvz on top of that: shouldn't the recent addition of "text-based custom editors"-API make this much less likely to happen? If an extension operates on the same file that the text editor does, then I would assume that the extension is NOT introducing a new working copy because it uses the text-document based API which surely does not register a working copy.

And then, if an extension does NOT work on text documents because the underlying resource is binary, I would not think that the workbench could register a working copy because the workbench cannot open binary files anyway.

Please clarify in what situation we would hit this.

**Comment by mjbvz** [comment]
Yes, text based custom editors solve some of these problems. However users can still cause two working copies to be registered in a few different ways, and extensions can also trigger this if they don't understand how we expected them to behavior

And after debugging through this again, I think there's also a bug:

1. Open two image editor for the same resource side-by-side
1. Now reopen one of them using VS Code's standard editor

The re-open should open the standard binary editor. However, during the reopen, we try to register a working copy [while constructing a `TextFileEditorModel`](https://github.com/microsoft/vscode/blob/01dbe67aa9a079424f5500c16296f4355f7b354d/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L107) for the binary resource. This happens before we detect the resource is binary and discard the `TextFileEditorModel`.

Should we move the call to `registerWorkingCopy` into `TextFileEditorModel .load` instead?

**Comment by bpasero** [comment]
Can you extract that into a separate issue?

**Comment by bpasero** [comment]
> However users can still cause two working copies to be registered in a few different ways

How?

**Comment by mjbvz** [comment]
89d3b684a288ef65b8d4eaf47a92262fdbe8e72b added a custom scheme that should now prevent this
